File: frontend/src/api/sessions.ts
import { apiGet, apiPost, apiPut } from './client'

export interface SessionSummary {
  id: string
  name: string
  device_id: string | null
  started_at: string
  ended_at: string | null
  status: 'recording' | 'completed'
  packet_count: number
  metric_count: number
  event_count: number
}

export interface SessionPacket {
  device_id: string
  raw: string
  ts: string
}

export interface SessionMetric {
  metric_name: string
  value: number
  unit?: string
  ts: string
}

export interface SessionEvent {
  type: string
  message?: string
  ts: string
}

export interface SessionDetail {
  id: string
  device_id: string | null
  name: string
  started_at: string
  ended_at: string | null
  packets: SessionPacket[]
  metrics: SessionMetric[]
  events: SessionEvent[]
}

export async function listSessions(): Promise<{ sessions: SessionSummary[] }> {
  return apiGet<{ sessions: SessionSummary[] }>('/api/sessions/')
}

export async function getSession(id: string): Promise<SessionDetail> {
  return apiGet<SessionDetail>(`/api/sessions/${id}`)
}

export async function getSessionInfo(id: string): Promise<SessionSummary> {
  return apiGet<SessionSummary>(`/api/sessions/${id}/info`)
}

export async function startSession(deviceId: string | null, name?: string): Promise<{ id: string; status: string }> {
  return apiPost('/api/sessions/', { device_id: deviceId, name })
}

export async function stopSession(id: string): Promise<{ id: string; status: string }> {
  return apiPost(`/api/sessions/${id}/stop`, {})
}

export async function renameSession(id: string, name: string): Promise<any> {
  return apiPut(`/api/sessions/${id}/rename`, { name })
}

export async function getSessionPackets(id: string, limit = 1000): Promise<SessionPacket[]> {
  return apiGet<SessionPacket[]>(`/api/sessions/${id}/packets?limit=${limit}`)
}

export async function getSessionMetrics(id: string, limit = 1000): Promise<SessionMetric[]> {
  return apiGet<SessionMetric[]>(`/api/sessions/${id}/metrics?limit=${limit}`)
}

export async function getSessionEvents(id: string, limit = 1000): Promise<SessionEvent[]> {
  return apiGet<SessionEvent[]>(`/api/sessions/${id}/events?limit=${limit}`)
}

File: backend/app/api/routes/sessions.py
"""Session management API routes."""
from fastapi import APIRouter, HTTPException
from typing import Optional, List

from app.core.session_recorder import session_recorder
from app.models import SessionCreate, SessionResponse

router = APIRouter(prefix="/sessions", tags=["sessions"])


@router.post("/", response_model=dict)
async def create_session(create: SessionCreate):
    """Start a new recording session."""
    import uuid
    session_id = str(uuid.uuid4())
    await session_recorder.start_session(
        session_id,
        device_id=create.device_id,
        name=create.name,
    )
    return {"id": session_id, "status": "recording"}


@router.post("/{session_id}/stop")
async def stop_session(session_id: str):
    """Stop a recording session."""
    await session_recorder.stop_session(session_id)
    return {"id": session_id, "status": "stopped"}


@router.put("/{session_id}/rename")
async def rename_session(session_id: str, body: dict = None):
    """Rename a session. Accepts {"name": "..."} in body."""
    name = body.get("name", "") if body else ""
    if not name:
        raise HTTPException(status_code=400, detail="name is required")
    success = session_recorder.rename_session(session_id, name)
    if not success:
        raise HTTPException(status_code=404, detail="Session not found")
    return {"id": session_id, "name": name, "status": "renamed"}


@router.get("/{session_id}")
async def get_session(session_id: str):
    """Get session details with all data."""
    session = await session_recorder.load_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session


@router.get("/{session_id}/info")
async def get_session_info(session_id: str):
    """Get session metadata (lightweight)."""
    info = session_recorder.get_session_info(session_id)
    if not info:
        raise HTTPException(status_code=404, detail="Session not found")
    return info


@router.get("/", response_model=List[dict])
async def list_sessions():
    """List all sessions."""
    return session_recorder.list_sessions()


@router.get("/{session_id}/packets")
async def get_session_packets(session_id: str, limit: int = 1000):
    """Get packets from a session."""
    session = await session_recorder.load_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.get("packets", [])[-limit:]


@router.get("/{session_id}/metrics")
async def get_session_metrics(session_id: str, limit: int = 1000):
    """Get metrics from a session."""
    session = await session_recorder.load_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.get("metrics", [])[-limit:]


@router.get("/{session_id}/events")
async def get_session_events(session_id: str, limit: int = 1000):
    """Get events from a session."""
    session = await session_recorder.load_session(session_id)
    if not session:
        raise HTTPException(status_code=404, detail="Session not found")
    return session.get("events", [])[-limit:]
